### ISSUE TITLE
Investigate MS project on Triton (https://github.com/microsoft/triton-shared)

### ISSUE THREAD

**Issue by etiotto** [intel/intel-xpu-backend-for-triton#165]
The goal is to understand the project at https://github.com/microsoft/triton-shared and determine how it fits within our Triton port.

**Comment by sommerlukas** [comment]
We have investigated the project mentioned above in our team. 

The general idea of the project is to function as a bridge between Triton and backends that are not specific for Triton. The entry point that the project chose to bridge to other backends is the `Linalg` dialect.

By bridging between Triton and Linalg, triton-shared aims to enable more re-use of backends. For example, using triton-shared as a bridge would allow a single, Linalg-based backend to be shared between Triton and other frontends (e.g., `torch-mlir`).

The bridge currently only works in one direction, i.e., it can convert from Triton IR to Linalg and other "core" MLIR dialects, but it does not support conversion back into the Triton world.

triton-shared currently works in three steps and treats the pointer portion and the value portion of a Triton program differently. While the aim for the pointer world, i.e., everything related to `tt.ptr`, load, stores etc., is to convert it to `memref`, the value world is translated to Linalg operations working on tensor-typed values.

The three steps involved are;
1. Triton to Structured: An elaborate analysis analyses Triton pointers and the address calculations, with the aim to identify structured accesses. All accesses detected as structured are converted to operations from the triton-shared `tts` dialect. All operations for which the detection failed will remain as TritonIR operations for now.
2. Triton to Linalg: As described above, Triton operations on values, e.g., arithmetic operations are converted to Linalg (e.g., `generic`) operations
3. Structured to Memref: The operations detected and converted to `tts` in the first step are now converted to `memref` operations. All remaining Triton operations are treated as scalar operations with `affine` loads/stores and loops. 

Next to investigating the project code and MLIR produced by triton-shared, we also tested triton-shared with the Triton code generated by TorchInductor for five models that were used in the TPP-MLIR evaluation (https://github.com/plaidml/mlir-generator/tree/main/pytorch/torch-dynamo/models). 

The investigation revealed a number of things:
* The analysis for structured memory accesses doesn't yet cover all the cases in Triton generated by TorchInductor. For four out of the five models, the memory access analysis wasn't able to detect the access as structured (the fifth failed for other reasons). It seems that some of the Triton code generated by TorchInductor was not yet covered by the analysis implementation, showing language coverage gaps in the analysis.
* The fifth model failed due to some currently unsupported Triton language feature not yet covered by triton-shared, again showing gaps in language coverage.
* In the structured-to-memref conversion, triton-shared generates allocations and copies. According to the maintainers, this is to make sure that the memory underpinning the memref/tensor is contiguous and should be optimized away by optimization passes that currently not part of the triton-shared reference compilation pipeline for CPU. We also raised this [here](https://github.com/microsoft/triton-shared/discussions/126). 

In particular the first point shows that the complex analysis required to convert from Triton to Linalg & Memref can cause issues, as it needs to cover all language constructs in order to operate effectively. Whether this approach of converting to a rather high-level MLIR dialect based on a complex analysis is really idiomatic in the MLIR framework is an open question.
